### ISSUE TITLE
Bare asserts in Hail

### ISSUE THREAD

**Issue by poneill** [hail-is/hail#12952]
### What happened?

**A Motivating Example**

The statement:

`hl.experimental.load_dataset("gnomad_pca_variant_loadings", version='2.1', reference_genome='GRCh38')`

raises the following error:

```
AssertionError                            Traceback (most recent call last)
Cell In[32], line 1
----> 1 hl.experimental.load_dataset("gnomad_pca_variant_loadings", version='2.1', reference_genome='GRCh38')

File ~/.local/lib/python3.8/site-packages/hail/experimental/datasets.py:115, in load_dataset(name, version, reference_genome, region, cloud)
    107     raise ValueError(f'Region {repr(region)} not available for dataset'
    108                      f' {repr(name)} on cloud platform {repr(cloud)}.\n'
    109                      f'Available regions: {regions}.')
    111 path = [dataset['url'][cloud][region]
    112         for dataset in datasets[name]['versions']
    113         if all([dataset['version'] == version,
    114                 dataset['reference_genome'] == reference_genome])]
--> 115 assert len(path) == 1
    116 path = path[0]
    117 if path.startswith('s3://'):

AssertionError: 
```

I'm a new Hail user and don't have the full context here, but it seems like there are at least three problems:

1.  An assert failed in production code, which indicates either the presence of a bug or an incorrect use of assert (e.g. using assert to check for value errors).
2. The assert has no corresponding error message, so the user learns that something has gone wrong but can't easily tell what.
3.  The assert is bare.  Bare asserts can get optimized out of code in ways that are difficult to foresee in advance, and are generally deprecated in favor of the `if error_condition: raise AssertionError(...)` pattern (see: https://discuss.python.org/t/stop-ignoring-asserts-when-running-in-optimized-mode/13132).

**The Big Picture**

The bare assert pattern is used over 3k times in Hail.  To be fair, many of these usages occur in test directories, where they're fine.  But they also occur in application code, and often in the dangerous form `assert(expr1, expr2)` which will never fail (because a tuple with two falsy elements is truthy in python).  These asserts are never actually getting checked.  

Fixing all of them would be a heavy lift.  One compromise solution might be to add a bare assert rule to the linter (e.g. https://pypi.org/project/flake8-assert-msg/).  This would prevent the introduction of further bare asserts to the codebase, and encourage authors to clean up existing bare asserts on files they touch.

The `assert` keyword is an unfortunate language wart that makes it very easy for developers to write error-checking code that is itself incorrect.  I'd encourage considering the alternate pattern `if error_condition: raise AssertionError(...)` and gradually migrating the codebase off of bare asserts.

Thanks for all of your work on Hail!

### Version

hail 0.2.115-10932c754edb

**Comment by danking** [comment]
Hey @poneill !

`-O`/`PYTHONOPTIMIZE` is explicitly defined as the "turn off asserts" option in [the docs](https://docs.python.org/3/using/cmdline.html#cmdoption-O). If you disable asserts, you'll get even more inscrutable errors. I recommend against doing that.

If you see any `assert(x, y)` in the code base, please file a PR or a bug. We'll fix it.

We will not replace asserts with if-raise.

---

As to the bug you've found: yes this is a bug in Hail. We incorrectly assume that if there is at least one dataset with the right version and at least one dataset with the right reference genome that there's a dataset with the right version *and* reference genome. That logic is obviously false. I'll have someone fix this in the next couple weeks.

As to the root issue: the Hail annotation database doesn't have a GRCh38 version of `gnomad_pca_variant_loadings` version 2.1. This is because [gnomAD](https://gnomad.broadinstitute.org/downloads#v2-liftover) hasn't published a GRCh38 version of their 2.1 variant loadings.

**Comment by poneill** [comment]
Hi @danking, thanks for this.

On the topic of asserts, there are really two interacting issues:

1.  Asserts are intended to be used in order to assure invariants, i.e. conditions that should always be true.  In correct code assertions should never raise, so disabling them should have no consequences at runtime.  In practice, however, they are often casually used to catch value errors, which can be expected to occur if a user-facing method receives bad/nonsensical inputs (e.g. here: https://github.com/hail-is/hail/blob/1940547d35ddddb084ad52684e36153c1e03a331/hail/python/hailtop/hailctl/dataproc/diagnose.py#L62)
2.  Python's language design allows anyone calling your code to disable asserts for optimization purposes, because disabling asserts should never change the semantics of the program.

Putting these two features together, you can arrive at a situation where a user thinks they're turning off asserts (which should never raise anyway) and instead stops catching value errors (whose absence can never be guaranteed).

All that said, if the final answer is: "if you invoke `-O` you deserve what's coming", I'm happy to drop it :)

Thanks for taking a look at the example.  If I understand you correctly, it sounds like I passed the wrong inputs to the function, in which case it might be clearer to raise a ValueError instead of an AssertionError in the end.

On a closer look, it seems like most of the instances of `assert(x, y)` are actually in scala code-- my mistake.

Thanks again for looking into this.
